### PR TITLE
User-friendly rnn sequencing

### LINKED ISSUE #21844: [RLlib] PPO cannot use standard optimizer when using recurrent policies
### Search before asking

- [x] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

This is a continuation of https://github.com/ray-project/ray/issues/19976, where RNN+PPO requires `simple_optimizer=True` to function. Since `simple_optimizer` is being deprecated, recurrent policies must be made to work with the new multi-GPU optimizer.

I will take a look at this and hopefully fix it shortly.

### Versions / Dependencies

`ray-1.9.2`

### Reproduction script

```python
import torch
import numpy as np
import gym
from typing import Union, Dict, List, Tuple, Any
import ray
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.utils.typing import ModelConfigDict, TensorType
from ray.rllib.policy.rnn_sequencing import add_time_dimension
from ray.tune import register_env
from ray.rllib.agents.ppo import PPOTrainer
from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole

# Pad to the correct size and crash
# or follow the rnn_sequencing code and don't crash
USE_CORRECT_SHAPE = False

class TestRNN(TorchModelV2, torch.nn.Module):
    def __init__(
        self,
        obs_space: gym.spaces.Space,
        action_space: gym.spaces.Space,
        num_outputs: int,
        model_config: ModelConfigDict,
        name: str,
        **custom_model_kwargs,
    ):
        TorchModelV2.__init__(
            self, obs_space, action_space, num_outputs, model_config, name
        )
        torch.nn.Module.__init__(self)
        self.num_outputs = num_outputs
        self.input_dim = gym.spaces.utils.flatdim(obs_space)
        self.act_space = action_space
        self.act_dim = gym.spaces.utils.flatdim(action_space)
        self.cur_val = None

        self.policy = torch.nn.Linear(self.input_dim, self.act_dim)
        self.vf = torch.nn.Linear(self.input_dim, 1)

    def get_initial_state(self):
        return [torch.zeros(0)]

    def value_function(self):
        assert self.cur_val is not None, "must call forward() first"
        return self.cur_val

    def forward(
        self,
        input_dict: Dict[str, TensorType],
        state: List[TensorType],
        seq_lens: TensorType,
    ) -> Tuple[TensorType, List[TensorType]]:

        flat = input_dict["obs_flat"]

        if USE_CORRECT_SHAPE:
            max_seq_len = seq_lens.max()
        else:
            # max_seq_len here is copied from rllib RNN code
            # see https://github.com/ray-project/ray/blob/2d24ef0d3234867ac329b10ae3a11b9b7119d17b/rllib/models/torch/recurrent_net.py#L75
            # but it doesn't make sense...
            # it should be max_seq_len = seq_len.max()
            max_seq_len = flat.shape[0] // seq_lens.shape[0]

        padded = add_time_dimension(
            flat,
            max_seq_len=max_seq_len,
            framework="torch",
            time_major=False
        )

        B = padded.shape[0]
        T = padded.shape[1]

        # If this fails, then we have "extra" padding in the RNN
        # We shouldn't need to pad the time dimension more than the longest
        # sequence
        if seq_lens.max() != T:
            print(f'seq_lens.max() is {seq_lens.max()} but input temporal dim is {T}')
            print(flat.reshape(B, T, -1) == 0)
            raise Exception('seq_len mismatch')

        flattened = padded.reshape(-1, padded.shape[-1])
        logits = self.policy(flattened)
        self.cur_val = self.vf(flattened).squeeze(1)
        state = state

        return logits, state

register_env(StatelessCartPole.__name__, StatelessCartPole)
MAX_SEQ_LEN = 200
CFG = {
    "env_config": {},
    "framework": "torch",
    "model": {
        "custom_model": TestRNN,
        "max_seq_len": MAX_SEQ_LEN,
    },
    "num_workers": 0,
    "num_gpus": 0,
    "env": StatelessCartPole,
    "horizon": MAX_SEQ_LEN,
    "sgd_minibatch_size": MAX_SEQ_LEN,
}
ray.init(object_store_memory=3e10)
analysis = ray.tune.run(
    PPOTrainer,
    config=CFG,
)
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by smorad**
Signed-off-by: Steven Morad <smorad@anyscale.com>

## Why are these changes needed?
There is a lot of mystery around how recurrent models work. Since we don't have time for a full redesign, we should at least try and hide the nasty bits from the user. Previously, it was not clear which `max_seq_len` should be passed to `add_time_dimension`: `SampleBatch.max_seq_len` or `SampleBatch['seq_lens'].max()` (neither of these, it's actually `SampleBatch['obs'].shape[0] // SampleBatch['seq_lens'].shape[0]`). This is because slicing the `SampleBatch` introduces unnecessary padding beyond the longest sequence in the slice (e.g. `seq_lens = [2,3,4]`, true shape: `obs.shape[3, 10, F]`. Fixing this would require implementing some unpadding functionality in `SampleBatch` or changing how the slice operator works, which seems like a breaking change.

This makes `add_time_dimension` more easily callable from the model forward pass, and reduces the chance of users introducing bugs. Now, `add_time_dimension` takes the full `seq_lens` tensor and calculates the dimensionality, rather than having the user calculate the dimensionality and pass it to `add_time_dimension` as `max_seq_len`.
## Related issue number

Closes https://github.com/ray-project/ray/issues/21844
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/rnn_sequencing.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def pad_batch_to_sequences_of_same_size(
 def add_time_dimension(
     padded_inputs: TensorType,
     *,
-    max_seq_len: int,
+    seq_lens: TensorType,
```
Comment (sven1977) [inline rllib/policy/rnn_sequencing.py:164]: Can you fix the docstring for this new arg?

**Review by sven1977 — APPROVED** [review]
Thanks @smorad , looks all good now.
